### ISSUE TITLE
Server side search is returning wrong results

### ISSUE THREAD

**Issue by splint33** [doxygen/doxygen#9839]
Hello, I have recently upgraded my Doxygen installation from 1.9.3 to 1.9.6.
I am using the server side search, in order to search into function brief & details in my Python project.

For example, I currently have the keyword "TWS" in 23 brief functions.

On Doxygen 1.9.3, when searching for "TWS", all 23 functions are showing up on the result page.
On Doxygen 1.9.6, when searching for "TWS", a completly unlinked function is showing up in the result page, with "Matches: tws(23)".

Searching "TWS" on Doxygen 1.9.3:
![image](https://user-images.githubusercontent.com/4028855/217312041-26d64cfe-b09f-4f54-99d4-6c97d2bdb59e.png)

Searching "TWS" on Doxygen 1.9.6:
![image](https://user-images.githubusercontent.com/4028855/217311138-9c4daed1-474e-4717-8846-318d449f5360.png)

**Comment by splint33** [comment]
I have crafted a little Stub project with four4 functions, three of them with the keyword "TWS" in their brief tag, and another one, completely unrelated, without TWS in the brief tag.

The doxyfile is located in the configuration folder, and I ran "doxygen at_ms.doxyfile" on the configuration folder in order to generate the documentation.

The documentation is then displayed using a web server.
You will need to change the OUTPUT_DIRECTORY, as it is pointing to a local network share.

As mentioned before, the search is working perfectly on Doxygen 1.9.3, and seems to be broken on 1.9.6.

When crafting the Stub project, I have found that the wrongly displayed results change based on its name.
I hope you can find & fix the issue, if you need more information, please let me know.

[StubProject.zip](https://github.com/doxygen/doxygen/files/10683437/StubProject.zip)

**Comment by albert-github** [comment]
Thanks for the small project.
The essential settings here are (I think)::
```
SEARCHENGINE           = YES
SERVER_BASED_SEARCH    = YES
EXTERNAL_SEARCH        = NO
```

I see that there are some changes, between the 1.9.3 version and the 1.9.6 version, though I cannot test it as I have no search server running.

@doxygen do you have an idea?

**Comment by splint33** [comment]
I have done more tests, it seems that the problem appeared in the version 1.9.4, and was not present in the version 1.9.3.

**Comment by albert-github** [comment]
@splint33 Interesting observation.

I did a diff between the results of the 1.9.3 and 1.9.4 versions and see:
```
diff -r html_1_9_3_server/menu.js html_1_9_4_server/menu.js
31c31,39
<         result+='<li><a href="'+relPath+data.children[i].url+'">'+
---
>         var url;
>         var link;
>         link = data.children[i].url;
>         if (link.substring(0,1)=='^') {
>           url = link.substring(1);
>         } else {
>           url = relPath+link;
>         }
>         result+='<li><a href="'+url+'">'+
Binary files html_1_9_3_server/search/search.idx and html_1_9_4_server/search/search.idx differ
```
(hopefully I didn't filter out relevant things).

I also checked the 1.9.4 `search.idx` against the 1.9.6 version and they are identical (looked at the md5sum of both files)

- the change in the `menu.js` doesn't look like to be relevant for the search problem
- the change of `search.idx` looks relevant to me
  -  what happens when you take the `search.idx` from the 1.9.3 version and place this in the 1.9.4 version?
      Interesting would be to know whether or not the overview would be OK again (the links might not be OK).
      This would give a bit of an insight in whether the problem lies in the `search.idx`

**Comment by splint33** [comment]
Replacing the search.idx from a 1.9.4 generated documentation by a  search.idx from a 1.9.3 generated documentation seems to resolve the issue.

**Comment by albert-github** [comment]
At first glance I didn't see any differences in the writing of the `search.idx` file, so it looks like it has to do with the contents of the of the fields in the `search.idx` file and the way these fields are generated.

Hopefully @doxygen has an idea.

**Comment by albert-github** [comment]
I'm doing some bisecting see that there are 2 changes in the `search.idx`, the first change looks harmless to me but to be on the safe size ...

- could you please see if the attached file [search.tar.gz](https://github.com/doxygen/doxygen/files/10686119/search.tar.gz) (unfortunately I had to pack it in a compressed tar file) works.

**Comment by splint33** [comment]
I have tested your search.idx file, and the issue is present

**Comment by albert-github** [comment]
Thanks for the very quick testing, the result was unexpected for me.

Looks like:
```
Commit: 1a921af825f3bdd890accd35a6896a92ddde41ec [1a921af]
Date: Thursday, January 13, 2022 7:37:55 PM
```
is still OK but
```
Commit: 30a6a1ac33b522ec5d00808b68b8f7428861df89 [30a6a1a]
Date: Friday, January 14, 2022 7:45:57 PM
```
is wrong.

I now undid that change in the master version and created a new `search.idx`
@splint33 can you please test this attached `seach.idx` as well: [search_1.tar.gz](https://github.com/doxygen/doxygen/files/10686257/search_1.tar.gz)

**Comment by splint33** [comment]
The last search.idx file is working fine !

**Comment by albert-github** [comment]
@splint33 Thanks again for the very fast testing, so I could do some further testing. I hope that @doxygen can shine a light on it.

The difference between the "correct" and "wrong" results is cause by the  call to `initSearchIndexer()` and subsequently some items land in the `search.idx`. In the small example the places I identified, in doxygen.cpp, are in the loop of: `for (const auto &fn : *Doxygen::inputNameLinkedMap)` and in the lambda function `auto addMemberToSearchIndex = [](const MemberDef *md)`.
{Probably there are more places in `addToIndices` but the call to `addToIndices` is now effective due to the place where `initSearchIndexer();` is called.

@doxygen 
- should the `initSearchIndexer()` already be active for the `Doxygen::searchIndex->kind()==SearchIndexIntf::Internal` in `addToIndices`?
- if so how to solve the problem in the `search.idx`?

**Comment by doxygen** [comment]
I think a found a better solution. The root of the problem was that when `setCurrentDoc()` was called multiple times for the same entry, the `m_urlIndex` was only set correctly the first time, causing additional words to not be logged under the correct search key for subsequent entries.

**Comment by albert-github** [comment]
@doxygen sounds as a better solution.

@splint33 I created a new idx file: [search.tar.gz](https://github.com/doxygen/doxygen/files/10716133/search.tar.gz), could you please test it?

**Comment by splint33** [comment]
The last search.tar.gz you have provided is working fine ! 
I'm glad you've managed to find & fix the issue, I hope it will be released soon !

**Comment by doxygen** [comment]
This issue was previously marked 'fixed but not released',
which means it should be fixed in doxygen version 1.9.7.
Please verify if this is indeed the case. Reopen the
issue if you think it is not fixed and please include any additional information
that you think can be relevant (preferably in the form of a self-contained example).
